### PR TITLE
@lefnire => Initial API support including specs

### PR DESCRIPTION
**Pull request by switz**
Don't merge yet. Going to clean up this PR first. I'll notify you when it's ready.

`x-api-key` and `x-api-user` should be sent as headers on all authorized requests.

To run tests, just run `npm install`. Then run `mocha test/api.mocha.coffee`. Let me know if any tests don't pass. 
## GET
- /api/v1/user - (get user status)
  - `x-api-user: uid`
  - `x-api-key: api token`
- /api/v1/user/tasks
  - `x-api-user: uid`
  - `x-api-key: api token`
  - `type: habit | daily | todo | reward` (optional)
- /api/v1/user/task/:id - (get task)
  - `x-api-user: uid`
  - `x-api-key: api token`
## POST
- /api/v1/user/task - (create new task)
  - `x-api-user: uid`
  - `x-api-key: api token`
  - `type: habit | daily | todo | reward` (required)
  - `title: title` (required)
  - `text: text` (required)
  - `value: value`
  - `note: note`
## PUT
- /api/v1/user/task/:id - (update task)
  - update title or text
  - `x-api-user: uid`
  - `x-api-key: api token`

### CODE REVIEW

**Comment by switz** [comment]
If you have any idea how we can account for the API not having sessions, please let me know. Racer needs some way of knowing which query is coming from the API. https://github.com/switz/habitrpg/blob/bc5abba179685effbb437c208643730fdd15f1e4/src/server/store.coffee#L35

**Comment by ruddfawcett** [comment]
:+1:

**Comment by ruddfawcett** [comment]
When do you think you'll have it cleaned up by?

**Comment by ruddfawcett** [comment]
@switz https://github.com/lefnire/habitrpg/issues/524

**Comment by switz** [comment]
For now we'll be using the header mechanism. If we decide that people want to authenticate, then we can redesign the API. Let's continue discussion about it in #524.  

If Tyler is okay with it, then this should be in place by tonight. Still a few more things to clean up and one bug to fix.

**Comment by switz** [comment]
@lefnire Once we solve the `store.coffee` [`return accept(true)`](https://github.com/lefnire/habitrpg/pull/522/files#L6R35) issue this should be ready to merge. Please read through https://github.com/lefnire/habitrpg/pull/522/files and let me know if **anything** is askew (just leave an inline note). Run `mocha test/api.mocha.coffee` to test the endpoints locally. If all the tests pass (and we've protected the api via `store.coffee`), merge away. This is far from complete – but, it should enable people to begin writing apps via the API.

I'm working on an auth endpoint, but running into some issues with regards to hashing the password. Let's pair on that whenever you have a second. https://github.com/switz/habitrpg/compare/switz:api...switz:api-auth

**Comment by ruddfawcett** [comment]
@switz soo excited!!! Ready to extend PHP whenever it's up!  If you need a tester, lemme know!

**Comment by lefnire** [comment]
@switz i'm gonna jump into this soon as I can, sorry about that. Stay tuned, and you rock.

**Comment by ruddfawcett** [comment]
Yeah you do!

**Comment by switz** [comment]
By the way, the server runs inline during the tests on port `1337`. So, no need to run your server separately while the tests are running (but if you are, they won't interfere).

**Comment by lefnire** [comment]
Nice!

File: package.json [hunk 1]
```diff
@@ -23,8 +23,11 @@
     "mongoskin": "*",
     "nconf": "*",
     "icalendar": "git://github.com/lefnire/node-icalendar#master",
+    "nodetime": "*",
```
Comment (switz) [inline package.json]: @lefnire not sure where this came from, might just be leftover from a bad merge. Is it still necessary?

**Comment by switz** [comment]
Just added a validation [middleware](https://github.com/switz/habitrpg/blob/c080ddc6832cad40e252497b421bacc4e268a5c5/src/server/api.coffee#L55) for `POST /user/task` and `PUT /task/:id`. There's a bit of logic added in exclusively for the `PUT` request, but that allows us to avoid sandwiching another middleware in between two anonymous functions. [Purely superficial](https://github.com/switz/habitrpg/blob/c080ddc6832cad40e252497b421bacc4e268a5c5/src/server/api.coffee#L79) coffeescript. The validation isn't perfect, but it gives us a nice base to work off of.

Just noticed this, but should we prepend `/user` to `PUT|GET /task/:id`? You should only really be able to update and get your own tasks, no?

**Comment by lefnire** [comment]
we should prepend `/user`/ because we'll be allowing cross-profile task assignment (see [this trello](https://trello.com/card/cross-profile-challenges-quests/50e5d3684fe3a7266b0036d6/58)) in the near future. nodetime is indeed a vestige, hmm I hope nothing else got lost in the merges - i'll have to step through some git trees.

**Comment by lefnire** [comment]
then eventually we'll want the party system, but we'll kick this around for now and come back later

**Pull request merged**